src/main/java/nextstep/subway/domain/Sections.java

```diff
@@ -13,23 +14,166 @@ public class Sections {
+			validateSectionDistance(newSection, sameDownStationSection.get());
+			addSectionDownStationBased(newSection, sameDownStationSection.get());
+			return;
+		}
+
+		if(isSameNewSectionDownStationAndUpStation(newSection)){
+			sectionList.add(newSection);
+			return;
+		}
+
+		if(isSameNewSectionUpStationAndDownStation(newSection)){
+			sectionList.add(newSection);
+			return;
+		}
+
+		throw new IllegalArgumentException("구간추가가 불가합니다.");
+	}
+
+	private void validateExisingStation(Section newSection) {
+		if(findSameUpAndDownStation(newSection)){
+			throw new IllegalArgumentException("상행역과 하행역이 모두 등록되어있습니다.");
+		}
+	}
+
+	public Optional<Section> getSameUpStationSection(Section newSection){
+		return this.sectionList.stream()
+			.filter(section -> section.isSameUpStation(newSection))
+			.findFirst();
+	}
+
+	private void validateSectionDistance(Section newSection, Section section) {
+		if(newSection.isLongerDistance(section)){
+			throw new IllegalArgumentException("추가하려는 구간의 길이는 기존 구간의 길이와 같거나 길수 없습니다.");
+		}
+	}
+
+	private Optional<Section> getSameDownStationSection(Section newSection) {
+		return this.sectionList.stream()
+			.filter(section -> section.isSameDownStation(newSection))
+			.findFirst();
+	}
+
+	private void addSectionDownStationBased(Section newSection, Section section) {
+		sectionList.remove(section);
+		sectionList.add(new Section(section.getLine(), section.getUpStation(), newSection.getUpStation(),
+			section.getDistance() - newSection.getDistance()));
+		sectionList.add(new Section(section.getLine(), newSection.getUpStation(), newSection.getDownStation(),
+			newSection.getDistance()));
+	}
+
+	private void addSectionUpStationBased(Section newSection, Section section) {
+		sectionList.remove(section);
+		sectionList.add(new Section(section.getLine(), section.getUpStation(), newSection.getDownStation(),
+			newSection.getDistance()));
+		sectionList.add(new Section(section.getLine(), newSection.getDownStation(), section.getDownStation(),
+			section.getDistance() - newSection.getDistance()));
+	}
+
+	public boolean isSameNewSectionDownStationAndUpStation(Section newSection) {
+		Section firstSection = getFirstSection();
+		return firstSection.getUpStation().equals(newSection.getDownStation());
+	}
+
+	public boolean isSameNewSectionUpStationAndDownStation(Section newSection) {
+		Section lastSection = getLastSection();
+		return lastSection.getDownStation().equals(newSection.getUpStation());
 	}
 
 	public List<Station> getStations() {
+		if(sectionList.isEmpty()){
+			return Collections.emptyList();
+		}
+
+		List<Station> sortedStationList = new ArrayList<>();
+		Section section = getFirstSection();
+		Section lastSection = getLastSection();
+
+		sortedStationList.add(section.getUpStation());
+
+		while(!section.equals(lastSection)){
+			sortedStationList.add(section.getDownStation());
+			section = nextSection(section);
+		}
+
+		sortedStationList.add(lastSection.getDownStation());
+		return sortedStationList;
```

Comment:
Stream API를 사용할 수도 있을 것 같아요. 대략 이런식은 어떨까요?

```suggestion
		List<Station> stations = sectionList.stream().map(Section::getUpStation).collect(toList());
		stations.add(.getLastSection().getDownStation());
		return stations;
```